src/main/scala/com/target/data_validator/ValidatorTable.scala

```diff
@@ -216,13 +239,78 @@ case class ValidatorHiveTable(
   override def substituteVariables(dict: VarSubstitution): ValidatorTable = {
     val newDb = getVarSub(db, "db", dict)
     val newTable = getVarSub(table, "table", dict)
-    val newKeyColumns = keyColumns.map(v => v.map(x => getVarSub(x, "keyCol", dict)))
-    val newCondition = condition.map(x => getVarSub(x, "condition", dict))
-    val newChecks = checks.map(_.substituteVariables(dict))
+    val newBases = substituteKeyColsCondsChecks(dict)
+
+    val ret = ValidatorHiveTable(
+      newDb, newTable,
+      newBases.keyColumns.map(_.toList),
+      newBases.condition,
+      newBases.checks
+    )
+    this.getEvents.foreach(ret.addEvent)
+    ret
+  }
+}
+
+/**
+ * Enables usage of the normal `spark.read.format(String).options(Map[String,String]).load()` fluent API
+ * for creating a [[DataFrame]] using a dynamic loader.
+ *
+ * For reading well-known formats such as Parquet or ORC,
+ * use [[ValidatorParquetFile]] or [[ValidatorOrcFile]], respectively.
+ * Under the hood, these just call `spark.read.format("orc")` but let's let Spark handle that.
+ *
+ * @param format the format that the [[DataFrameReader]] should use
+ * @param options options that the [[DataFrameReader]] should use
+ * @param loadData If present, this will passed to [[DataFrameReader.load(String)]],
+ *                 otherwise the DFR will use the parameterless version
+ */
+case class ValidatorSpecifiedFormatLoader(
+  format: String,
+  keyColumns: Option[List[String]],
+  condition: Option[String],
+  checks: List[ValidatorBase],
+  options: Option[Map[String, String]] = None,
+  loadData: Option[String] = None
```

Comment:
Is this really optional? Spark API boils down to a list of source string specs (sometimes empty for cases like kafka), might want to make the yaml accept a list instead?

Reply:
…I just went ahead with the move to a list. It was quick to implement and adjust tests.